### ISSUE TITLE
Ch. 05

### ISSUE THREAD

**Issue by jsung0228** [jsung0228/2023-Q1-DL-Book-Review#4]
### 오류 역전파법
---

**개요**
- Stochastic Gradient Descent 복습
- 단일 신경망에서의 Gradient Descent 적용
- 오차역전법을 통한 효과적인 training

**1. Stochastic Gradient Descent 복습**

아래의 그림과 같이, Gradient Descent를 통해서 model을 학습시킬 수 있다. 
![image](https://user-images.githubusercontent.com/59510027/230418760-72a809f4-7113-41b7-b872-32eb4179e164.png)

문제는, 전체 과정에 대해서 Gradient Descent를 실시하게 될 경우, 과정이 지나치게 느리다는 것이다. 
Cost 함수 J(t)에 대해서, 다음과 같이 쓸 수 있다. 

![image](https://user-images.githubusercontent.com/59510027/230423894-acadaaa7-6379-4dd8-a7bc-44d40fd431dd.png)

문제는 모든 pararmeter에 대해서, 이 전체 샘플값들을 다 넣기에는 너무 시간이 많이 걸린다. 
위의 식에서는 잘 안적혀 있지만, 실제로 gradient descent 를 하기 위해서는 cost function에 대해서 모든 값들을 다 집어넣고 나서 더해준 다음에 paramter에 대해서 training해줘야 하는데, 문제는 모든 데이터들을 forward pass 해야한다....

      **Stochastic Gradient Descent** : 그냥 random하게 일부를 뽑아서, 얘네들의 에러함수를 줄이는 방향으로 하자...

즉, 샘플링을 통해서, Gradient Descent를 계산할 수도 있다. 
이렇게 될 경우에는, Variance가 너무 커지는 문제가 있다. → Mini Batch Stochastic Gradient Descent
하나씩만 하는 것은 좀 그러니까, sum을 살리되, mini batch 에 대해서 실행하자의 컨셉

![image](https://user-images.githubusercontent.com/59510027/230419414-20520087-7bfa-4b36-a1d7-377e9e283627.png)

**2. 오차역전법에 대한 이해**

가장 간단한 single perceptron에 대해서 생각해보자
우리의 목적은 training하기 위해서는 cost 함수를 특정한 weight에 대해서 편미분하여서, w값을 조절해야한다. 

하지만, 편미분을 해서 구하려고 한다면 상당히 복잡한 식이 나올 것이다. 
      → 합성함수의 편미분을 활용하면, 훨씬 쉽게 구할 수 있다. 

![image](https://user-images.githubusercontent.com/59510027/230428743-68184893-2893-4bf9-859c-9bd5b57bdf16.png)

아래의 그림에서 알 수 있듯이, 어떤 weight가 cost 함수에 들어가기 위해서는 아래의 과정을 걸친다. 

Weight 함수 → 활성화 함수 → Loss 함수 계산

![image](https://user-images.githubusercontent.com/59510027/230429000-fc6b4048-7a74-4895-b42c-f2d9c9ae6652.png)

위의 식을 자세히 보게 되면, 결국 함수는 세 개의 구성요소가 있다. 

- Input : 알고 있는 값
- Output : Forward Path에서 저장되어 있는 값
- Target : 이미 알고 있는 값
     → Forward Path 를 계산할 때 썻던 값들의 사칙연산으로 구할 수 있다. (계산해야하는 횟수는 많다.)

![image](https://user-images.githubusercontent.com/59510027/230435894-7d379adc-f8c3-4aa4-b936-80ccedd48243.png)

Multi-Layer 에 대해서도 마찬가지로, input / output / target의 사칙연산으로 표현할 수 있다. 

![image](https://user-images.githubusercontent.com/59510027/230436825-995d398e-21fb-43b8-983c-52fc93b35994.png)

![KakaoTalk_20230408_124858242_01](https://user-images.githubusercontent.com/59510027/230701966-7f122556-41ca-4c64-a03f-d58114e27187.jpg)

![KakaoTalk_20230408_124858242](https://user-images.githubusercontent.com/59510027/230701969-144835e4-e900-4995-a819-2f8c229f6463.jpg)

![KakaoTalk_20230408_124852546](https://user-images.githubusercontent.com/59510027/230701975-8c15775b-c147-4a85-b791-5f9b633e3b48.jpg)

**Comment by jsung0228** [comment]
**밑바닥부터 오류역전파 모델의 구현**  

---

먼저, 아래의 그림을 보면서 우리가 구현하고자 하는 Neural Network 의 필요 요소에 대해서 이해해보자.

![image](https://user-images.githubusercontent.com/59510027/231778024-23d2853e-0f26-486d-a75b-a8b59c3371b6.png)

복습을 해보자면, 우리가 이미지 분석을 위해서 필요한 함수는 다음과 같다. 

- Softmax
- Affine Layer
- Activation (ReLU)

역전파 오류는 이미 메모리에 저장된 데이터 +  합성함수 규칙을 활용해서, gradient descent를 효율적으로 해줄 수 있다는 것을 배웠다. 

![image](https://user-images.githubusercontent.com/59510027/231803871-f829cf02-a2a2-48c6-9551-e3f53b2a7349.png)


실제로 파이썬에서는 효율적인 코드 구현을 위해서, 위의 세가지 요소에 forward와 backward 함수를 내장한다. 
각 함수의 예시를 보면서, forward 와 backward 함수를 어떻게 작성해야 하는지 알아보자.

---

**1. 활성화 함수** 

**ReLU 함수**는 아래와 같이 표현할 수 있다. 

- Forward : x<0 일 경우 0, x>0일 경우 x
- Backward : 미분값이 x<0 일 경우 0 이며, x>0 일 경우, 1이므로 상류 값을 toss

![image](https://user-images.githubusercontent.com/59510027/231804837-433da1d7-6863-4ec3-b734-2d957511031b.png)


```
# ReLU 계층
class Relu:
    def __init__(self):
        self.mask = None
 
    def forward(self, x):
        self.mask = (x <= 0)
        out = x.copy()
        out[self.mask] = 0
 
        return out
 
    def backward(self, dout):
        dout[self.mask] = 0
        dx = dout
 
        return dx
```

**Sigmoid**에 대해서는 아래와 같이 된다. 

![image](https://user-images.githubusercontent.com/59510027/231806048-67896693-f913-4d36-8bd5-3dc3766dbd2f.png)

![image](https://user-images.githubusercontent.com/59510027/231805836-0d1e7551-4784-4ece-9ad1-2a7bb7a0f28d.png)

```
# sigmoid 구현
class Sigmoid:
    def __init__(self):
        self.out = None
 
    def forward(self, x):
        out = sigmoid(x)
        self.out = out
        return out
 
    def backward(self, dout):
        dx = dout * (1.0 - self.out) * self.out
 
        return dx
```

**2. Affine 함수** 

Affine 함수는 Weight, 하류층, Bias로 이루어져있다. 
아래의 그림을 이해하기 위해서는, `matrix의 dot product에 대한 편미분 식을 이해해야할 것 같은데, 설명이 부족하다. `

![image](https://user-images.githubusercontent.com/59510027/231808556-c804c1eb-0719-4bb0-a7fb-a41827e2b8c8.png)

이를 코드로 구현하면 다음과 같다. 

```
# Affine 구현
class Affine:
  def __init__(self, W, b):
    self.W = W
    self.b = b
    self.x = None
    self.dW = None
    self.db = None
 
  def forward(self, x):
    self.x = x
    out = np.dot(x, self.W) + self.b
 
    return out
 
  def backward(self, dout):
    dx = np.dot(dout, self.W.T)
    self.dW = np.dot(self.x.T, dout)
    self.db = np.sum(dout, axis=0)
 
    return dx
```

**3. Softmax 함수**

`아래의 그림도 잘 이해가 되진 않는다. `

![image](https://user-images.githubusercontent.com/59510027/231810425-6f717a24-c4d8-4f4d-a3bb-3ff76147fea5.png)

`원핫벡터는 무엇일까??`

```
# Softmax-with-Loss 계층 구현
# common / layer.py
 
class SoftmaxWithLoss:
    def __init__(self):
        self.loss = None # 손실
        self.y = None # softmax의 출력
        self.t = None # 정답레이블 (원-핫 벡터)
 
    def forward(self, x, t):
        self.t = t
        self.y = softmax(x)
        self.loss = cross_entropy_error(self.y, self.t)
        
        return self.loss
 
    def backward(self, dout=1):
        batch_size = self.t.shape[0]
        if self.t.size == self.y.size: # 정답레이블이 원-핫 벡터일 경우
            dx = (self.y - self.t) / batch_size
        else:
            dx = self.y.copy()
            dx[np.arange(batch_size), self.t] -= 1
            dx = dx / batch_size
        
        return dx
```
-----

오류역전파 구현!!!

- 가설 : 신경망에는 적응 가능한 가중치와 편향이 있고, 이 가중치와 편향을 훈련 데이터에 적응하도록 조정하는 과정이 '학습'

- 미니배치 : 훈련 데이터 중 일부를 무작위로 가져옴 (= 미니배치)
- 기울기 산출 : 미니배치의 손실 함수 값을 줄이기 위해 각 가중치 매개변수의 기울기를 구함
- 매개변수 갱신 : 가중치 매개변수를 기울기 방향으로 아주 조금 갱신
- 반복 : 위의 단계를 반복해준다. 

`아래의 내용은 잘 이해가 되지 않는다....`

```
import sys, os
sys.path.append('/deep-learning-from-scratch')
import numpy as np
from common.layers import *
from common.gradient import numerical_gradient
from collections import OrderedDict
 
 
class TwoLayerNet:
 
    def __init__(self, input_size, hidden_size, output_size, weight_init_std = 0.01):
        # 가중치 초기화
        self.params = {}
        self.params['W1'] = weight_init_std * np.random.randn(input_size, hidden_size)
        self.params['b1'] = np.zeros(hidden_size)
        self.params['W2'] = weight_init_std * np.random.randn(hidden_size, output_size) 
        self.params['b2'] = np.zeros(output_size)
 
        # 계층 생성
        # OrderedDict = 순서가 있는 딕셔너리, 순서 기억
        # 순전파 때는 계층을 추가한 순서대로 / 역전파 때는 계층 반대 순서로 호출
        self.layers = OrderedDict()
        self.layers['Affine1'] = Affine(self.params['W1'], self.params['b1'])
        self.layers['Relu1'] = Relu()
        self.layers['Affine2'] = Affine(self.params['W2'], self.params['b2'])
 
        self.lastLayer = SoftmaxWithLoss()
        
    def predict(self, x):
        for layer in self.layers.values():
            x = layer.forward(x)
        
        return x
        
    # x: 입력데이터, t : 정답레이블
    def loss(self, x, t):
        y = self.predict(x)
        return self.lastLayer.forward(y, t)
    
    def accuracy(self, x, t):
        y = self.predict(x)
        y = np.argmax(y, axis=1)
        if t.ndim != 1 : t = np.argmax(t, axis=1)
        
        accuracy = np.sum(y == t) / float(x.shape[0])
        return accuracy
        
    # x: 입력데이터, t : 정답레이블
    def numerical_gradient(self, x, t):
        loss_W = lambda W: self.loss(x, t)
        
        grads = {}
        grads['W1'] = numerical_gradient(loss_W, self.params['W1'])
        grads['b1'] = numerical_gradient(loss_W, self.params['b1'])
        grads['W2'] = numerical_gradient(loss_W, self.params['W2'])
        grads['b2'] = numerical_gradient(loss_W, self.params['b2'])
        
        return grads
        
    def gradient(self, x, t):
        # forward, 순전파
        self.loss(x, t)
 
        # backward, 역전파
        dout = 1
        dout = self.lastLayer.backward(dout)
        
        layers = list(self.layers.values())
        layers.reverse()
        for layer in layers:
            dout = layer.backward(dout)
 
        # 결과 저장
        grads = {}
        grads['W1'], grads['b1'] = self.layers['Affine1'].dW, self.layers['Affine1'].db
        grads['W2'], grads['b2'] = self.layers['Affine2'].dW, self.layers['Affine2'].db
 
        return grads
```
